app/cdap/components/PipelineDiff/util/reactflowGraph.ts

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright © 2023 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+import { getGraphLayout } from 'components/hydrator/helpers/DAGhelpers';
+import { getPluginIcon, objectQuery } from 'services/helpers';
+import { IPipelineConfig, IPipelineDiffMap, IPipelineStage } from '../types';
+import { MarkerType } from 'reactflow';
+import { PLUGIN_TYPES } from 'components/hydrator/components/Canvas/constants';
+import {
+  generatePluginMapKey,
+  getCustomIconSrc,
+  pluginDiffColors,
+  shouldShowAlertsPort,
+  shouldShowErrorsPort,
+} from './helpers';
+import { getConnectionDiffKey, getStageDiffKey } from './diff';
+
+/**
+ * Given a pipeline configuration, constructs a pipeline graph that can be displayed
+ * in a reacflow canvas.
+ * @param config a pipeline's configuration
+ * @returns nodes and connections to be displayed in a reactflow pipeline graph
+ */
+export function getReactflowPipelineGraph(
+  config: IPipelineConfig | null,
+  availablePluginsMap: any,
+  diffMap: IPipelineDiffMap
+) {
+  if (!config) {
+    return { nodes: [], connections: [] };
+  }
+  const graphNodes = config.stages.map((stage) => {
+    return {
+      name: stage.name,
+      type: stage.plugin.type,
+      label: stage.plugin.label,
+    };
+  });
+  const graphConnections = config.connections;
+  const graph = getGraphLayout(graphNodes, graphConnections, 200);
+
+  const stageNameToStage: { [stageName: string]: IPipelineStage } = {};
+
+  const nodes = config.stages.map((stage) => {
+    stageNameToStage[stage.name] = stage;
+    const pluginMapKey = generatePluginMapKey(stage);
+
+    const hasErrorAndAlert =
+      shouldShowAlertsPort(availablePluginsMap, pluginMapKey) ||
```

Comment:
can you combine this into one function in your helper file?

Reply:
combine the type line into this helper funciton as well?

Reply:
done.